### PR TITLE
Fresh

### PR DESCRIPTION
**Pull request by pzel**
NOTE: This is a work in progress.

# F-F-Fresh users for each new test!

There is nothing preventing the developer from running all their escalus stories in parallel .... except the fact that the set of users is very restricted, and they are used in all tests cases.

In short, there is a data dependency, and shared user state on the server.

`escalus_fresh` solves these problems by taking an existing `config` proplist, modifying usernames with a suffix unique to the given story, and creating these users. These users are sure to have *no state* waiting for them on the server.

## How do I use this?

  *  Roll up all your test groups into one. Really, test case names should be descriptive on their own. Take the opportunity to remove dead comments and name things for what they are.

```erlang
groups() ->
  [{errors, [badfoo, badbar]},
   {frobs, [tired_test, happy_test]}].
```
turns into
```erlang
groups() ->
  [{all_tests, [sending_badfoo_raises_error, 
                sending_badbar_raises_error,
                a_frob_frobs_when_tired,
                a_frob_frobs_when_happy]}].
```
 *  Add `[parallel, shuffle]` to your group. Run it and watch it fail with login errors.
 *  Change instances of `escalus:story` to `escalus:fresh_story` and watch your tests run in parallel with fresh, non-conflicting users created for every test case.

## What if I need access to the user specs in the `Config` proplist?

In an ideal world, all user data would be available as part of the logged-in `#client{}`s API. Can you make it happen? If so, then there's no problem! If you can't, use the following hack:

Change:
```erlang
escalus:story(Config, [{alice, 1}, {bob, 1}, fun(Alice, Bob) ->
    D = get_some_user_data(Config, bob), %% OUCH! bob's data may have been shadowed!
    use_data(D),
    ...
```
into:

```erlang
escalus:fresh_story_with_config(Config, [{alice,1},{bob,1}], fun(FreshConfig, Alice, Bob) ->
    D = get_some_user_data(FreshConfig, bob), %% this **IS** the bob you're looking for!
    use_data(D),
    ...
```

# How can I clean up after the autogenerated users?

I believe that running your escalus stories with fresh users should provide you with the framework and mindset that **you don't have to care** about server state generated by and for **other users**. The whole idea behind this move is to **untangle data dependencies** and move tests towards self-contained, property-like behavior. And make things faster, of course.

Anyway.

`escalus_fresh:clean()` will clean up all fresh users generated globally. Use this in `end_per_SUITE` to satisfy the gods of server purity checks.

### CODE REVIEW

File: src/escalus_fresh.erl [hunk 1]
```diff
@@ -0,0 +1,38 @@
+-module(escalus_fresh).
+-export([story/3, story_with_config/3]).
+
+story_with_config(Config, UserSpecs, StoryFun) ->
+    FreshSpecs = fresh_specs(Config, UserSpecs, fresh_suffix()),
+    case length(FreshSpecs) == length(UserSpecs) of
```
Comment (gadgetci) [inline src/escalus_fresh.erl]: According to **Elvis**:
> The code in the following (LINE, COL) locations has the same structure: (6, 5), (14, 5).

File: src/escalus_fresh.erl [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    {N, lists:keyreplace(username, 1, UserConfig, {username, NewName})}.
+
+select(UserResources, FullSpecs) ->
+    Fst = fun({A,_}) -> A end,
+    UserNames = lists:map(Fst, UserResources),
+    lists:filter(fun({Name, _}) -> lists:member(Name, UserNames) end,
+                 FullSpecs).
+
+fresh_suffix() ->
+    {_,S,US} = erlang:now(),
```
Comment (gadgetci) [inline src/escalus_fresh.erl]: According to **Elvis**:
> Missing space after "," on line 36

File: src/escalus_fresh.erl [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    [ make_fresh_username(Spec, StorySuffix)
+      || Spec <-select(TestedUsers, AllSpecs) ].
+
+make_fresh_username({N, UserConfig}, Suffix) ->
+    {username, OldName} = proplists:lookup(username, UserConfig),
+    NewName = << OldName/binary, Suffix/binary >>,
+    {N, lists:keyreplace(username, 1, UserConfig, {username, NewName})}.
+
+select(UserResources, FullSpecs) ->
+    Fst = fun({A,_}) -> A end,
```
Comment (gadgetci) [inline src/escalus_fresh.erl]: According to **Elvis**:
> Missing space after "," on line 30

File: src/escalus_fresh.erl (hunk 1 above)
Comment (michalwski) [inline src/escalus_fresh.erl]: That's good one @gadgetci!

File: src/escalus_fresh.erl [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        false -> ct:fail("failed to get required users"); _ -> ok end,
+    FreshConfig = escalus_users:create_users(Config, FreshSpecs),
+    escalus:story(FreshConfig, UserSpecs,
+                  fun(Args) -> apply(StoryFun, [FreshConfig|Args]) end).
+
+story(Config, UserSpecs, StoryFun) ->
+    FreshSpecs = fresh_specs(Config, UserSpecs, fresh_suffix()),
+    case length(FreshSpecs) == length(UserSpecs) of
+        false -> ct:fail("failed to get required users"); _ -> ok end,
+    FreshConfig = escalus_users:create_users(Config, FreshSpecs),
```
Comment (michalwski) [inline src/escalus_fresh.erl]: Where are these users removed from the server?

Comment (pzel) [inline src/escalus_fresh.erl]: Never.  No subsequent tests will ever reference these users.

File: test/stream_automation_SUITE.erl [hunk 5]
```diff
@@ -105,6 +105,7 @@ simple_ack(Config) ->
 %% them in escalus stories.
 ack_after_four_messages(Config) ->
     escalus_ejabberd:rpc(mod_stream_management, set_ack_freq, [4]),
+    escalus_ejabberd:rpc(ejabberd_config, add_global_option, [{shaper,normal,global}, {maxrate, 10000}]),
```
Comment (gadgetci) [inline test/stream_automation_SUITE.erl]: According to **Elvis**:
> Missing space after "," on line 108

Comment (gadgetci) [inline test/stream_automation_SUITE.erl]: According to **Elvis**:
> Line 108 is too long:     escalus_ejabberd:rpc(ejabberd_config, add_global_option, [{shaper,normal,global}, {maxrate, 10000}]),.

File: test/escalus_fresh_SUITE.erl [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                  {password, <<"p">>}]}
+          ]}] ++ empty_config(),
+    given_escalus_started(C),
+
+    % when
+    escalus_fresh:create_users(C, [{ada, 1}]),
+
+    % then
+    [{Suffix,
+      [{escalus_users,[{ada,[{username,<<"ada" , Suffix/binary>>},
```
Comment (gadgetci) [inline test/escalus_fresh_SUITE.erl]: According to **Elvis**:
> Missing space after "," on line 56

File: test/escalus_fresh_SUITE.erl [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    ].
+
+it_creates_a_global_table_on_startup(_) ->
+    given_escalus_started(empty_config()),
+    %% then
+    [] = ets:tab2list(internal_fresh_db()).
+
+it_registers_fresh_user_specs_in_table(_) ->
+    C = [{escalus_users,
+          [{al, [{username, <<"a">>}, {server, <<"0.0.0.0">>}, {password, <<"p">>}]}]}]
```
Comment (gadgetci) [inline test/escalus_fresh_SUITE.erl]: According to **Elvis**:
> The code in the following (LINE, COL) locations has the same structure: (18, 12), (31, 12).

File: test/escalus_fresh_SUITE.erl [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    given_escalus_started(C),
+
+    % when
+    escalus_fresh:create_users(C, [{ada, 1}]),
+
+    % then
+    [{Suffix,
+      [{escalus_users,[{ada,[{username,<<"ada" , Suffix/binary>>},
+                             {server,<<"0.0.0.0">>},
+                             {password,<<"p">>}]}]},
```
Comment (gadgetci) [inline test/escalus_fresh_SUITE.erl]: According to **Elvis**:
> Missing space after "," on line 58

File: test/escalus_fresh_SUITE.erl [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+          ]}] ++ empty_config(),
+    given_escalus_started(C),
+
+    % when
+    escalus_fresh:create_users(C, [{ada, 1}]),
+
+    % then
+    [{Suffix,
+      [{escalus_users,[{ada,[{username,<<"ada" , Suffix/binary>>},
+                             {server,<<"0.0.0.0">>},
```
Comment (gadgetci) [inline test/escalus_fresh_SUITE.erl]: According to **Elvis**:
> Missing space after "," on line 57

File: test/escalus_fresh_SUITE.erl [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+    C = [{escalus_users,
+          [{al, [{username, <<"a">>}, {server, <<"0.0.0.0">>}, {password, <<"p">>}]}]}]
+        ++ empty_config(),
+    given_escalus_started(C),
+
+    % when
+    catch(escalus_fresh:story(C, [{al, 1}], fun id/1)),
+
+    % then
+    [{Suffix, [{escalus_users, [{al, [{username, <<"a", Suffix/binary>>}, _, _]}]},_]}]
```
Comment (gadgetci) [inline test/escalus_fresh_SUITE.erl]: According to **Elvis**:
> Missing space after "," on line 26

File: src/escalus_fresh.erl [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+
+%%% Internals
+nasty_global_table() -> escalus_fresh_db.
+
+ensure_table_present(T) ->
+    case ets:info(T) of
+        undefined ->
+            P = spawn(fun() -> ets:new(T, [named_table, public]),
+                               receive bye -> ok end end),
```
Comment (gadgetci) [inline src/escalus_fresh.erl]: According to **Elvis**:
> The expression on line 69 and column 32 is nested beyond the maximum level of 3.

File: src/escalus_fresh.erl [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    lists:filter(fun({Name, _}) -> lists:member(Name, UserNames) end,
+                 FullSpecs).
+
+fresh_suffix() ->
+    {_, S, US} = erlang:now(),
+    L = lists:flatten([integer_to_list(S rem 100), ".", integer_to_list(US)]),
+    list_to_binary(L).
+
+
+%%
```
Comment (gadgetci) [inline src/escalus_fresh.erl]: According to **Elvis**:
> Line 97 has 1 trailing whitespace characters.

File: src/escalus_fresh.erl [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+delete_users({_Suffix, Conf}) ->
+    Plist = proplists:get_value(escalus_users, Conf),
+    {ok,_} = escalus_users:delete_users(Conf, Plist),
+    ok.
+
+ensure_table_present(T) ->
+    case ets:info(T) of
+        undefined ->
+            P = spawn(fun() -> ets:new(T, [named_table, public]),
+                               receive bye -> ok end end),
```
Comment (gadgetci) [inline src/escalus_fresh.erl]: According to **Elvis**:
> The expression on line 70 and column 32 is nested beyond the maximum level of 3.

File: src/escalus_fresh.erl [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+    [ok|_] = pmap(fun delete_users/1, ets:tab2list(nasty_global_table())),
+    ets:delete_all_objects(nasty_global_table()),
+    ok.
+
+%%% Internals
+nasty_global_table() -> escalus_fresh_db.
+
+delete_users({_Suffix, Conf}) ->
+    Plist = proplists:get_value(escalus_users, Conf),
+    {ok,_} = escalus_users:delete_users(Conf, Plist),
```
Comment (gadgetci) [inline src/escalus_fresh.erl]: According to **Elvis**:
> Missing space after "," on line 63

File: test/escalus_fresh_SUITE.erl [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+                  {password, <<"p">>}]}
+          ]}] ++ empty_config(),
+    given_escalus_started(C),
+
+    % when
+    escalus_fresh:create_users(C, [{ada, 1}]),
+
+    % then
+    [{Suffix,
+      [{escalus_users,
```
Comment (gadgetci) [inline test/escalus_fresh_SUITE.erl]: According to **Elvis**:
> Line 56 has 1 trailing whitespace characters.

File: test/escalus_fresh_SUITE.erl [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+    ].
+
+it_creates_a_global_table_on_startup(_) ->
+    given_escalus_started(empty_config()),
+    %% then
+    [] = ets:tab2list(internal_fresh_db()).
+
+it_registers_fresh_user_specs_in_table(_) ->
+    C = [{escalus_users,
+          [{al, [{username, <<"A">>}, {server, <<"0.0.0.0">>}, {password, <<"p">>}]}]}]
```
Comment (gadgetci) [inline test/escalus_fresh_SUITE.erl]: According to **Elvis**:
> The code in the following (LINE, COL) locations has the same structure: (18, 39), (31, 39), (32, 39), (33, 40), (46, 19), (58, 17).

**Pull request merged**